util/chunk/checksum.go

```diff
@@ -0,0 +1,104 @@
+package chunk
+
+import (
+	"bufio"
+	"errors"
+	"hash/crc32"
+	"io"
+	"os"
+)
+
+const (
+	checksumBlockSize   = 1024
+	checksumSize        = 4
+	checksumPayloadSize = checksumBlockSize - checksumSize
+)
+
+var emptyChecksumBlock = make([]byte, checksumBlockSize)
+
+type checksum struct {
+	disk      *os.File
+	bufWriter *bufio.Writer
+}
+
+func newChecksum(disk *os.File) *checksum {
+	cs := &checksum{disk: disk}
+	cs.bufWriter = bufio.NewWriterSize(disk, checksumPayloadSize)
+	return cs
+}
+
+func (cks *checksum) Write(p []byte) (n int, err error) {
+	start, end := 0, len(p)
+	if end > checksumPayloadSize {
+		end = checksumPayloadSize
+	}
+	payload := p[start:end]
+	buf := make([]byte, checksumBlockSize)
+	for len(payload) > 0 {
+		// We need to fill in payload before calculating the checksum.
+		copy(buf[4:], payload)
+		checksum := crc32.Checksum(buf[4:], crc32.MakeTable(crc32.IEEE))
+		buf[0] = byte(checksum & 0xff000000 >> 24)
+		buf[1] = byte(checksum & 0x00ff0000 >> 16)
+		buf[2] = byte(checksum & 0x0000ff00 >> 8)
+		buf[3] = byte(checksum & 0x000000ff)
+		n1, err := cks.bufWriter.Write(buf)
+		n += n1
+		if err != nil {
+			return n, err
+		}
```

Comment:
IMO,  checksum unit is Block, so if the n1 != len(buf), it should generate an error.....